Add config flow to mystrom

pail23 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

The YAML configuration will be deprecated.


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Add config flow for mystrom switches and bulbs


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [x] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bdraco commented:
Looks like there is a conflict that needs to be resolved

bdraco commented on homeassistant/components/mystrom/__init__.py:
```diff
@@ -1 +1,82 @@
+            raise MyStromConnectionError()
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the myStrom component."""
+    hass.data[DOMAIN] = {}
+    return True
+
```
```suggestion
```

bdraco commented on homeassistant/components/mystrom/__init__.py:
```diff
@@ -1 +1,82 @@
+        raise PlatformNotReady() from err
+
+    if device:
+        hass.data[DOMAIN][entry.entry_id] = {"device": device, "info": info}
```
```suggestion
        hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {"device": device, "info": info}
```

bdraco commented on homeassistant/components/mystrom/__init__.py:
```diff
@@ -1 +1,82 @@
+    host = entry.data[CONF_HOST]
+    device = None
+    try:
+        info = await get_device_info(host)
+        device_type = info["type"]
+        if device_type in [101, 106, 107]:
+            device = MyStromSwitch(host)
+            await device.get_state()
+        elif device_type == 102:
+            mac = info["mac"]
+            device = MyStromBulb(host, mac)
+            await device.get_state()
+            if device.bulb_type not in ["rgblamp", "strip"]:
+                _LOGGER.error(
+                    "Device %s (%s) is not a myStrom bulb nor myStrom LED Strip",
+                    host,
+                    mac,
+                )
+                device = None
```
Please move code out of the try block that isn't expected to raise

pail23 replied:
I would appreciate a little hint on how to improve this one: `get_device_info` and the two `device.get_state` can both raise the `MyStromConnectionError`. I could wrap the 3 calls with a try block but this would look more complex to me. How can I best move code out of the try block without making the code more complex?

epenet replied:
It feels to me like maybe it should simply be:
```python
    try:
        info = await get_device_info(host)
    except MyStromConnectionError as err:
        _LOGGER.error("No route to myStrom plug: %s", host)
        raise ConfigEntryNotReady() from err

    device_type = info["type"]
    if device_type in [101, 106, 107]:
        device = MyStromSwitch(host)
    elif device_type == 102:
        mac = info["mac"]
        device = MyStromBulb(host, mac)
    else:
        # not sure what can be done here... maybe log error and return false?
        return False

    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
        "device": device,
        "info": info,
    }
    hass.config_entries.async_setup_platforms(entry, PLATFORMS)

    return True
```

All the other checks should probably be inside the individual platforms.

epenet replied:
If you really feel that `device.get_state()` is required, then you need wrap just that call in a second and third `try...except`.

pail23 replied:
Implemented as proposed

bdraco commented on homeassistant/components/mystrom/__init__.py:
```diff
@@ -1 +1,82 @@
+
+    except MyStromConnectionError as err:
+        _LOGGER.error("No route to myStrom plug: %s", host)
+        raise PlatformNotReady() from err
```
`PlatformNotReady` won't work here.  You'll need `ConfigEntryNotReady` instead

pail23 replied:
Changed to ConfigEntryNotReady

OneCyrus commented:
would be great if this PR doesn't die and gets a review again.

phedoreanu commented:
@bdraco what about another review and a merge? 😁

fabaff commented on homeassistant/components/mystrom/__init__.py:
```diff
@@ -1 +1,78 @@
+async def get_device_info(host: str) -> dict[str, Any]:
+    """Get the device info of a myStrom device."""
+    async with aiohttp.ClientSession() as session:
+        async with session.get(f"http://{ host}/api/v1/info") as response:
```
Newer devices are using a different endpoint. See https://github.com/home-assistant-ecosystem/python-mystrom/pull/34

pail23 replied:
I'm not sure whether I get your point: In the referenced PR the endpoint is `api/v1/info`. In the [mystrom documentation](https://api.mystrom.ch/#f37a4be7-0233-4d93-915e-c6f92656f129) the endpoint is `api/v1/info`.

What is the specific change you propose?

fabaff replied:
Depending on the firmware there are two different endpoints available to get the details about the switch. 

- Older firmware: `info.json`
- New firmware: `api/v1/info`

pail23 replied:
Implemented both versions

epenet commented on homeassistant/components/mystrom/config_flow.py:
```diff
@@ -0,0 +1,84 @@
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id="user", data_schema=STEP_USER_DATA_SCHEMA
+            )
```
To avoid duplicate code, we usually recommend that you invert the logic:
```suggestion
        errors = {}
        if user_input is not None:
            ...
        return self.async_show_form(
                step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors
            )
```

pail23 replied:
Implemented

epenet commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,75 @@
+DEVICE_NAME = "test-myStrom Device"
+
+
+async def test_form(hass: HomeAssistant) -> None:
```
I suggest that you create a mock_setup_entry fixture in a local `conftest.py`

```suggestion
pytestmark = pytest.mark.usefixtures("mock_setup_entry")


async def test_form(hass: HomeAssistant, mock_setup_entry: AsyncMock) -> None:
```

pail23 replied:
implemented

pail23 replied:
Now it should be in

pail23 committed:
rebase

pail23 committed:
fixed review comments

pail23 commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,75 @@
+    with patch(
+        "homeassistant.components.mystrom.async_setup_entry",
+        return_value=True,
+    ) as mock_setup_entry:
```
Implemented

pail23 commented on homeassistant/components/mystrom/translations/en.json:
```diff
@@ -0,0 +1,18 @@
+{
```
File removed

pail23 committed:
fix test

pail23 committed:
Update tests

pail23 committed:
Merge branch 'dev' into mystrom_config

pail23 committed:
Merge branch 'dev' into mystrom_config

pail23 committed:
Merge branch 'dev' into mystrom_config

pail23 committed:
increase test coverage

pail23 committed:
Merge branch 'dev' into mystrom_config

epenet commented on homeassistant/components/mystrom/config_flow.py:
```diff
@@ -0,0 +1,80 @@
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): str,
+        vol.Required("host"): str,
```
```suggestion
        vol.Required(CONF_HOST): str,
```

pail23 replied:
Implemented

epenet commented on homeassistant/components/mystrom/config_flow.py:
```diff
@@ -0,0 +1,80 @@
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): str,
```
You shouldn't need to ask for a name. It should come from the device info, or use a default entry title.
```suggestion
```

pail23 replied:
Removed

epenet commented on homeassistant/components/mystrom/config_flow.py:
```diff
@@ -0,0 +1,80 @@
+    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
+    """
+
+    if not host_valid(data[CONF_HOST]):
+        raise CannotConnect()
+
+    return {"title": data[CONF_NAME]}
```
Instead of this, you should actually try to get the device info. Then you can actually use the device info to set a unique-id, and maybe set a title.

epenet commented on homeassistant/components/mystrom/light.py:
```diff
@@ -121,6 +138,16 @@ def is_on(self):
+    def device_info(self) -> DeviceInfo:
+        """Return the device info for the light entity."""
+        return DeviceInfo(
+            identifiers={(DOMAIN, self._name)},
```
This is not a valid identifier.

pail23 replied:
I now use the mac address. Would this work?

epenet commented on homeassistant/components/mystrom/strings.json:
```diff
@@ -0,0 +1,18 @@
+      "unknown": "[%key:common::config_flow::error::unknown%]"
+    },
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
```
This is currently unused - but it should be implemented.

pail23 replied:
Implemented

epenet commented on homeassistant/components/mystrom/switch.py:
```diff
@@ -102,3 +119,13 @@ async def async_update(self) -> None:
+    def device_info(self) -> DeviceInfo:
+        """Return the device info for the light entity."""
+        return DeviceInfo(
+            identifiers={(DOMAIN, self._name)},
```
This is not a valid identifier.

pail23 replied:
changed to mac address

epenet commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,77 @@
+    assert len(mock_switch.mock_calls) == 1
+
+
+async def test_form_invalid_host(hass: HomeAssistant) -> None:
```
You should combine this with the test above, to have end-to-end flow:
init => FORM
configure with failure => FORM
configure success => CREATE_ENTRY

pail23 replied:
I have combed it test_form_combined

epenet commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,77 @@
+    assert result2["errors"] == {"base": "cannot_connect"}
+
+
+async def test_step_import(hass: HomeAssistant) -> None:
```
You should have a test to ensure duplicates are not created again.

pail23 replied:
Implemented in test_form_combined

pail23 committed:
implement some review comments

pail23 committed:
Merge branch 'dev' into mystrom_config

pail23 committed:
Merge branch 'dev' into mystrom_config

pail23 committed:
Merge branch 'dev' into mystrom_config

pail23 committed:
Enhance device check for old FWs and add tests

pail23 committed:
Merge branch 'dev' into mystrom_config

pail23 committed:
Merge branch 'dev' into mystrom_config

epenet commented on homeassistant/components/mystrom/__init__.py:
```diff
@@ -1 +1,79 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+async def get_device_info(host: str) -> dict[str, Any]:
+    """Get the device info of a myStrom device."""
+    async with aiohttp.ClientSession() as session:
+        async with session.get(f"http://{host}/api/v1/info") as response:
+            if response.status != 200:
+                async with session.get(f"http://{host}/info.json") as response:
+                    if response.status != 200:
+                        raise MyStromConnectionError()
+            content_type = response.headers.get("Content-Type", "")
+            if "application/json" in content_type:
+                return await response.json()
+            raise MyStromConnectionError()
```
I'm thinking that this whole `get_device_info` should maybe be a helper method directly in the library.
We shouldn't be making direct URL hits I think.

pail23 replied:
I created a [PR ](https://github.com/home-assistant-ecosystem/python-mystrom/pull/40) in the repository of the library...

pail23 replied:
@fabaff Do you have any recommendation on how to proceed with this?

pail23 replied:
`get_device_info` is now part of the library

pail23 committed:
Reworked exception handling

pail23 committed:
small code optimizations

pail23 committed:
fix test

pail23 committed:
Increase test coverage

pail23 commented:
@epenet All your comments should be resolved except the one regarding moving `get_device_info `to the library. There, a PR on the library is pending. Is there anything else I need to do before this PR can be merged?

pail23 committed:
Merge branch 'dev' into mystrom_config

pail23 committed:
Merge branch 'dev' into mystrom_config

pail23 committed:
Merge branch 'dev' into mystrom_config

pail23 committed:
Merge branch 'dev' into mystrom_config

pail23 committed:
Update __init__.py

changed from hass.config_entries.async_setup_platforms(entry, PLATFORMS) to await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

pail23 committed:
Merge branch 'dev' into mystrom_config

pail23 committed:
Update __init__.py

remove spaces

pail23 committed:
Merge branch 'dev' into mystrom_config

pail23 committed:
Merge branch 'dev' into mystrom_config

pail23 committed:
Merge branch 'dev' into mystrom_config

pail23 committed:
Merge branch 'dev' into mystrom_config

pail23 committed:
Merge branch 'home-assistant:dev' into mystrom_config

pail23 committed:
Bump python-mystrom to 2.2.0

pail23 committed:
Migrate to get_device_info from python-mystrom

pail23 committed:
Migrate to get_device_info from python-mystrom

pail23 committed:
Update __init__.py

pail23 committed:
update requirements_all.txt

pail23 committed:
update config_flow

pail23 committed:
fix tests

epenet commented on homeassistant/components/mystrom/manifest.json:
```diff
@@ -2,9 +2,10 @@
   "iot_class": "local_polling",
   "loggers": ["pymystrom"],
-  "requirements": ["python-mystrom==1.1.2"]
+  "requirements": ["python-mystrom==2.2.0"]
```
The library bump should be a preliminary PR.

pail23 replied:
I created the PR #93389 #93389

pail23 committed:
Merge branch 'dev' into mystrom_config

epenet commented on homeassistant/components/mystrom/light.py:
```diff
@@ -42,6 +47,23 @@
+
+    if device_type == 102:
+        device = hass.data[DOMAIN][entry.entry_id]["device"]
+        try:
+            await device.get_state()
+        except MyStromConnectionError as err:
+            _LOGGER.error("No route to myStrom bulb: %s", info["ip"])
+            raise ConfigEntryNotReady() from err
```
You should move this check to `__init__` instead.
You can raise `ConfigEntryNotReady` there, and when you reach the platform you don't need to worry about it anymore.

epenet commented on homeassistant/components/mystrom/light.py:
```diff
@@ -89,12 +109,12 @@ def __init__(self, bulb, name, mac):
     def unique_id(self):
         """Return a unique ID."""
-        return self._mac
+        return format_mac(self._mac)
```
Is this not a breaking change?
Why format it now, when it wasn't formatted previously?

epenet commented on homeassistant/components/mystrom/light.py:
```diff
@@ -121,6 +141,16 @@ def is_on(self):
+    def device_info(self) -> DeviceInfo:
+        """Return the device info for the light entity."""
+        return DeviceInfo(
+            identifiers={(DOMAIN, format_mac(self._mac))},
```
Note: I think it's OK to format it in the device info, but I think it shouldn't be formatted for the unique_id unless you implement a migration.

pail23 replied:
changed to mac

epenet commented on homeassistant/components/mystrom/switch.py:
```diff
@@ -28,14 +33,30 @@
+
+    if device_type in [101, 106, 107]:
+        device = hass.data[DOMAIN][entry.entry_id]["device"]
+        try:
+            await device.get_state()
+        except MyStromConnectionError as err:
+            _LOGGER.error("No route to myStrom plug: %s", info["ip"])
+            raise ConfigEntryNotReady() from err
```
Same as light: move this to `__init__`.

pail23 replied:
I moved the code to `__init__`

epenet commented on homeassistant/components/mystrom/switch.py:
```diff
@@ -71,7 +91,7 @@ def is_on(self):
     def unique_id(self):
         """Return a unique ID."""
-        return self.plug._mac  # pylint: disable=protected-access
+        return format_mac(self.plug.mac)
```
Same as light => this should not be formatted.

pail23 replied:
removed the fromat

epenet commented on homeassistant/components/mystrom/switch.py:
```diff
@@ -28,14 +33,30 @@
+        async_add_entities([MyStromSwitch(device)])
+
+
 async def async_setup_platform(
```
`async_setup_platform` should not create entities.
It should instead trigger an import and raise a `deprecated_yaml` issue in the issue registry.

pail23 replied:
Implemented

epenet commented on homeassistant/components/mystrom/light.py:
```diff
@@ -42,6 +47,23 @@
+        async_add_entities([MyStromLight(device, info["mac"])])
+
+
 async def async_setup_platform(
```
`async_setup_platform` should not create entities.
It should instead trigger an import and raise a `deprecated_yaml` issue in the issue registry.

pail23 replied:
Implemented

pail23 committed:
Update homeassistant/components/mystrom/__init__.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

pail23 committed:
Merge branch 'dev' into mystrom_config

pail23 commented on homeassistant/components/mystrom/__init__.py:
```diff
@@ -1 +1,64 @@
+
+from .const import DOMAIN
+
+PLATFORMS: list[Platform] = [Platform.SWITCH]
```
Good catch. I add Platform.LIGHT

pail23 commented on homeassistant/components/mystrom/config_flow.py:
```diff
@@ -0,0 +1,86 @@
+)
+
+
+def host_valid(host):
+    """Return True if hostname or IP address is valid."""
+    try:
+        if ipaddress.ip_address(host).version == (4 or 6):
+            return True
+    except ValueError:
+        disallowed = re.compile(r"[^a-zA-Z\d\-]")
+        return all(x and not disallowed.search(x) for x in host.split("."))
```
Removed

pail23 commented on homeassistant/components/mystrom/config_flow.py:
```diff
@@ -0,0 +1,86 @@
+    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
+    """
+
+    if not host_valid(data[CONF_HOST]):
+        raise CannotConnect()
```
Removed

pail23 commented on homeassistant/components/mystrom/config_flow.py:
```diff
@@ -0,0 +1,86 @@
+            else:
+                await self.async_set_unique_id(info["mac"])
+                self._abort_if_unique_id_configured()
+                return self.async_create_entry(title=info["mac"], data=user_input)
```
Unfortunatelly, there is no other information from the device api available except the mac address and the ip address. In a previous very, I had implemented a configurable name (which is also the case before this PR: The name is set in the configuration.yaml).
Would it make sense to implement the name config option again? From my perspective, this would be the most user friendly option given that such information is not available in the api. @epenet What do you think?

epenet replied:
I guess you can ask for a name, and store that in the entry title.
You shouldn't store it in the config data.
You can then use the entry title for naming devices and entities.

pail23 replied:
I gave it a try and implemented a title. Would this work?

pail23 committed:
Implemented review changes

pail23 committed:
increase test coverage

epenet commented on homeassistant/components/mystrom/light.py:
```diff
@@ -89,7 +100,7 @@ def __init__(self, bulb, name, mac):
     @property
     def name(self):
         """Return the display name of this light."""
-        return self._name
+        return self._mac
```
You probably think I'm being a pain... but I think in a preliminary PR you should move most of light and switch properties to attributes in `__init__` method.
For example:
```python
self._attr_name = name
self._attr_unique_id = mac
self._attr_brightness = ...
self._attr_hs_color = ...
self._attr_available = ...
self._attr_effect_list = ...
self._attr_is_on = ...
```

pail23 replied:
I submitted the PR #93591

epenet commented on homeassistant/components/mystrom/switch.py:
```diff
@@ -28,40 +31,54 @@
         self.relay = None
 
     @property
     def name(self):
         """Return the name of the switch."""
-        return self._name
+        return self.plug.mac
```
You probably think I'm being a pain... but I think in a preliminary PR you should move most of light and switch properties to attributes in `__init__` method.
For example:
```python
self._attr_name = name
self._attr_unique_id = plug.mac
self._attr_available = ...
self._attr_is_on = ...
```

pail23 replied:
I submitted the PR #93591

epenet commented on homeassistant/components/mystrom/switch.py:
```diff
@@ -102,3 +119,13 @@ async def async_update(self) -> None:
                 self._available = False
                 _LOGGER.error("No route to myStrom plug")
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return the device info for the light entity."""
+        return DeviceInfo(
+            identifiers={(DOMAIN, self.plug.mac)},
+            name=self.name,
+            manufacturer=ATTR_MANUFACTURER,
+            sw_version=self.plug.firmware,
+        )
```
This can be added as `self._attr_device_info` in `__init__` method.

pail23 replied:
done

epenet commented on homeassistant/components/mystrom/light.py:
```diff
@@ -121,6 +132,16 @@ def is_on(self):
         """Return true if light is on."""
         return self._state
 
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return the device info for the light entity."""
+        return DeviceInfo(
+            identifiers={(DOMAIN, self._mac)},
+            name=self.name,
+            manufacturer=ATTR_MANUFACTURER,
+            sw_version=self._bulb.firmware,
+        )
+
```
This can be added as `self._attr_device_info` in `__init__` method.

pail23 replied:
done

epenet reviewed (changes requested):
Please adjust the PR description to add a breaking change section, indicating the YAML deprecation.

pail23 committed:
Implemented user defined title

pail23 committed:
implemented user defined title

pail23 committed:
Additional review comments

pail23 commented on homeassistant/components/mystrom/__init__.py:
```diff
@@ -1 +1,79 @@
+    """Data class for mystrom device data."""
+
+    device: MyStromSwitch | MyStromBulb
+    info: dict
```
done

pail23 committed:
fix test

pail23 committed:
Merge branch 'dev' into mystrom_config

pail23 committed:
fix linter

pail23 committed:
fix linter

pail23 committed:
Merge branch 'dev' into mystrom_config

pail23 committed:
Merge branch 'dev' into mystrom_config

epenet commented on homeassistant/components/mystrom/models.py:
```diff
@@ -0,0 +1,13 @@
+    """Data class for mystrom device data."""
+
+    device: MyStromSwitch | MyStromBulb
+    info: dict
```
Can this be more specific?
```suggestion
    info: dict[str, Any]
```

epenet commented on homeassistant/components/mystrom/switch.py:
```diff
@@ -28,34 +31,55 @@
     """Representation of a myStrom switch/plug."""
 
-    def __init__(self, plug, name):
+    def __init__(self, plug, name) -> None:
```
Please avoid partial type hints.
This one (and the same in light) could be done in a separate (parallel) PR.

pail23 replied:
removed type hints (for this PR)

epenet commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,129 @@
+        assert result2["title"] == DEVICE_NAME
+        assert result2["data"] == {"host": "1.1.1.1"}
+
+    # test for duplicates
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
```
This should be a separate test.
```suggestion
async def test_form_combined(hass: HomeAssistant, mock_setup_entry: AsyncMock) -> None:
    """Test abort on duplicate."""
    entry = MockConfigEntry(domain=DOMAIN, data={CONF_HOST: "1.1.1.1"}, unique_id=DEVICE_MAC)
    entry.add_to_hass(hass)
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": config_entries.SOURCE_USER}
    )
```

pail23 replied:
implemented

epenet commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,129 @@
+    assert len(mock_setup_entry.mock_calls) == 1
+
+
+async def test_form_duplicates(
+    hass: HomeAssistant, mock_setup_entry: AsyncMock
+) -> None:
+    """Test we get the form."""
```
If you apply the changes above, I think this one can be removed.

pail23 replied:
Removed

epenet commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,129 @@
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": config_entries.SOURCE_USER}, data=conf
+        )
+        assert result["type"] == FlowResultType.FORM
```
Please ensure that this tests continues until CREATE or ABORT.

epenet commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,129 @@
+        "pymystrom.get_device_info",
+        side_effect=MyStromConnectionError(),
+    ):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": config_entries.SOURCE_USER}, data=conf
+        )
```
`SOURCE_USER` should have data `None` in `async_init`, followed by `async_configure`

pail23 replied:
Corrected

epenet commented on homeassistant/components/mystrom/__init__.py:
```diff
@@ -1 +1,71 @@
+        device=device,
+        info=info,
+    )
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
```
Maybe you should have two platform constants instead, and use a local variable:
```python
PLATFORMS_SWITCH = [Platform.SWITCH]
PLATFORMS_BULB = [Platform.LIGHT]
...
if device_type in [101, 106, 107]:
    platforms = PLATFORMS_SWITCH
elif device_type == 102:
    platforms = PLATFORMS_BULB
```

pail23 committed:
Update homeassistant/components/mystrom/models.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

pail23 committed:
fix review comments

pail23 committed:
fix missing import

pail23 committed:
simplify

epenet commented on homeassistant/components/mystrom/__init__.py:
```diff
@@ -1 +1,75 @@
+
+PLATFORMS_SWITCH = [Platform.SWITCH]
+PLATFORMS_BULB = [Platform.LIGHT]
+PLATFORMS = [Platform.SWITCH, Platform.LIGHT]
```
```suggestion
```

epenet commented on homeassistant/components/mystrom/__init__.py:
```diff
@@ -1 +1,75 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        hass.data[DOMAIN].pop(entry.entry_id)
```
```suggestion
    device_type = hass.data[DOMAIN][entry.entry_id].info["type"]
    if device_type in [101, 106, 107]:
        platforms = PLATFORMS_SWITCH
    elif device_type == 102:
        platforms = PLATFORMS_BULB
    if unload_ok := await hass.config_entries.async_unload_platforms(entry, platforms):
        hass.data[DOMAIN].pop(entry.entry_id)
```

epenet commented on homeassistant/components/mystrom/config_flow.py:
```diff
@@ -0,0 +1,76 @@
+
+        if user_input is not None:
+            try:
+                info = await validate_input(self.hass, user_input)
+            except CannotConnect:
```
```suggestion
                info = await pymystrom.get_device_info(self.hass, user_input[CONF_HOST])
            except MyStromConnectionError:
```

epenet commented on homeassistant/components/mystrom/config_flow.py:
```diff
@@ -0,0 +1,76 @@
+)
+
+
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
+    """Validate the user input allows us to connect.
+
+    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
+    """
+    try:
+        info = await pymystrom.get_device_info(data[CONF_HOST])
+    except MyStromConnectionError as error:
+        raise CannotConnect() from error
+
+    return {"mac": info.get("mac")}
```
This is not needed. It can be handled inline further down.
```suggestion
```

epenet commented on homeassistant/components/mystrom/config_flow.py:
```diff
@@ -0,0 +1,76 @@
+        return self.async_show_form(step_id="user", data_schema=schema, errors=errors)
+
+
+class CannotConnect(HomeAssistantError):
+    """Error to indicate we cannot connect."""
```
Not needed
```suggestion
```

epenet commented on homeassistant/components/mystrom/const.py:
```diff
@@ -1,2 +1,4 @@
 """Constants for the myStrom integration."""
 DOMAIN = "mystrom"
+DEFAULT_NAME = "myStrom"
+ATTR_MANUFACTURER = "myStrom"
```
```suggestion
MANUFACTURER = "myStrom"
```

pail23 replied:
done

epenet commented on homeassistant/components/mystrom/light.py:
```diff
@@ -17,13 +16,17 @@
+from homeassistant.helpers.issue_registry import IssueSeverity, async_create_issue
 from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType
 
+from .const import ATTR_MANUFACTURER, DOMAIN
```
```suggestion
from .const import DOMAIN, MANUFACTURER
```

pail23 replied:
done

epenet commented on homeassistant/components/mystrom/light.py:
```diff
@@ -40,30 +43,39 @@
+) -> None:
+    """Set up the myStrom entities."""
+    info = hass.data[DOMAIN][entry.entry_id].info
+    device_type = info["type"]
+
+    if device_type == 102:
+        device = hass.data[DOMAIN][entry.entry_id].device
+        async_add_entities([MyStromLight(device, entry.title, info["mac"])])
```
```suggestion
    device = hass.data[DOMAIN][entry.entry_id].device
    async_add_entities([MyStromLight(device, entry.title, info["mac"])])
```

epenet commented on homeassistant/components/mystrom/switch.py:
```diff
@@ -28,24 +31,39 @@
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the myStrom entities."""
+    info = hass.data[DOMAIN][entry.entry_id].info
+    device_type = info["type"]
+
+    if device_type in [101, 106, 107]:
+        device = hass.data[DOMAIN][entry.entry_id].device
+        async_add_entities([MyStromSwitch(device, entry.title)])
```
```suggestion
    device = hass.data[DOMAIN][entry.entry_id].device
    async_add_entities([MyStromSwitch(device, entry.title)])
```

epenet commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,133 @@
+        await hass.async_block_till_done()
+
+        assert result2["type"] == FlowResultType.CREATE_ENTRY
+        assert result2["title"] == DEVICE_NAME
```
It's better to be explicit in the assert.
```suggestion
        assert result2["title"] == "myStrom Device"
```

pail23 replied:
done

epenet commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,133 @@
+        )
+        await hass.async_block_till_done()
+
+        assert result2["type"] == FlowResultType.ABORT
```
Add an assert for the reason

epenet commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,133 @@
+        assert result2["type"] == FlowResultType.ABORT
+
+    assert len(mock_session.mock_calls) == 1
+    # assert len(mock_setup_entry.mock_calls) == 1
```
```suggestion
```

epenet commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,133 @@
+            DOMAIN, context={"source": config_entries.SOURCE_IMPORT}, data=conf
+        )
+        assert result["type"] == FlowResultType.CREATE_ENTRY
+        assert result["title"] == DEVICE_NAME
```
```suggestion
        assert result["title"] == "myStrom Device"
```

epenet commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,133 @@
+        await hass.async_block_till_done()
+
+        assert result2["type"] == FlowResultType.CREATE_ENTRY
+        # assert result2["title"] == DEVICE_NAME
+        # assert result2["data"] == {"host": "1.1.1.1"}
```
```suggestion
        assert result2["title"] == DEVICE_NAME
        assert result2["data"] == {"host": "1.1.1.1"}
```

epenet commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,133 @@
+        )
+        await hass.async_block_till_done()
+
+        assert result2["type"] == FlowResultType.FORM
```
Add an assert for the step

epenet commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,133 @@
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == FlowResultType.FORM
```
Add an assert for the step

pail23 replied:
I have added `assert result["step_id"] == "user"`. Is this what you meant?

epenet commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,133 @@
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == FlowResultType.FORM
```
Add an assert for the step

epenet commented on tests/components/mystrom/test_init.py:
```diff
@@ -0,0 +1,171 @@
+DEVICE_MAC = "6001940376EB"
+
+
+async def init_integration(hass, device_type, bulb_type="strip"):
```
```suggestion
async def init_integration(hass: HomeAssistant, config_entry: MockConfigEntry, device_type: int, bulb_type: str ="strip") -> None:
```

epenet commented on tests/components/mystrom/test_init.py:
```diff
@@ -0,0 +1,171 @@
+        new_callable=PropertyMock,
+        return_value=DEVICE_MAC,
+    ):
+        config_entry = MockConfigEntry(
+            domain=DOMAIN,
+            entry_id=ENTRY_ID,
+            unique_id="uuid",
+            data={CONF_HOST: "1.1.1.1"},
+            title=DEVICE_NAME,
+        )
+        config_entry.add_to_hass(hass)
```
I would make this a common fixture in `conftest.py`:
```suggestion
@pytest.fixture
def config_entry(hass: HomeAssistant) -> MockConfigEntry:
    """Create and add a config entry."""
    config_entry = MockConfigEntry(
        domain=DOMAIN,
        entry_id=ENTRY_ID,
        unique_id="uuid",
        data={CONF_HOST: "1.1.1.1"},
        title=DEVICE_NAME,
    )
    config_entry.add_to_hass(hass)
    return config_entry
```

pail23 replied:
done

epenet commented on tests/components/mystrom/test_init.py:
```diff
@@ -0,0 +1,171 @@
+    assert config_entry.state == ConfigEntryState.LOADED
+
+
+async def test_init_switch(hass: HomeAssistant) -> None:
+    """Test the initialization of a myStrom switch."""
+    await init_integration(hass, 101)
```
```suggestion
async def test_init_switch(hass: HomeAssistant, config_entry: MockConfigEntry) -> None:
    """Test the initialization of a myStrom switch."""
    await init_integration(hass, config_entry, 101)
```

epenet commented on tests/components/mystrom/test_init.py:
```diff
@@ -0,0 +1,171 @@
+    assert state is not None
+
+
+async def test_init_bulb(hass: HomeAssistant) -> None:
+    """Test the initialization of a myStrom bulb."""
+    await init_integration(hass, 102)
```
```suggestion
async def test_init_bulb(hass: HomeAssistant, config_entry: MockConfigEntry) -> None:
    """Test the initialization of a myStrom bulb."""
    await init_integration(hass, config_entry, 102)
```

epenet commented on tests/components/mystrom/test_init.py:
```diff
@@ -0,0 +1,171 @@
+    assert state is not None
+
+
+async def test_init_of_unknown_bulb(hass: HomeAssistant) -> None:
```
```suggestion
async def test_init_of_unknown_bulb(hass: HomeAssistant, config_entry: MockConfigEntry) -> None:
```

epenet commented on tests/components/mystrom/test_init.py:
```diff
@@ -0,0 +1,171 @@
+        new_callable=PropertyMock,
+        return_value=DEVICE_MAC,
+    ):
+        config_entry = MockConfigEntry(
+            domain=DOMAIN,
+            entry_id=ENTRY_ID,
+            unique_id="uuid",
+            data={CONF_HOST: "1.1.1.1"},
+            title="myStrom",
+        )
+        config_entry.add_to_hass(hass)
+
```
```suggestion
```

epenet commented on tests/components/mystrom/test_init.py:
```diff
@@ -0,0 +1,171 @@
+        assert config_entry.state == ConfigEntryState.SETUP_ERROR
+
+
+async def test_init_of_unknown_device(hass: HomeAssistant) -> None:
```
```suggestion
async def test_init_of_unknown_device(hass: HomeAssistant, config_entry: MockConfigEntry) -> None:
```

pail23 replied:
done

epenet commented on tests/components/mystrom/test_init.py:
```diff
@@ -0,0 +1,171 @@
+        "pymystrom.get_device_info",
+        side_effect=AsyncMock(return_value={"type": 103, "mac": DEVICE_MAC}),
+    ):
+        config_entry = MockConfigEntry(
+            domain=DOMAIN,
+            entry_id=ENTRY_ID,
+            unique_id="uuid",
+            data={CONF_HOST: "1.1.1.1"},
+            title="myStrom",
+        )
+        config_entry.add_to_hass(hass)
+
```
```suggestion
```

epenet commented on tests/components/mystrom/test_init.py:
```diff
@@ -0,0 +1,171 @@
+    assert config_entry.state == ConfigEntryState.SETUP_ERROR
+
+
+async def test_unload(hass: HomeAssistant) -> None:
+    """Test the unloading of a myStrom witch."""
+    await init_integration(hass, 101)
+    await hass.config_entries.async_unload(ENTRY_ID)
+    await hass.async_block_till_done()
+    state = hass.states.get("switch.test")
+    assert state is None
```
I think you should merge this with `test_init_switch` and rename it `test_init_switch_and_unload`

pail23 replied:
done

epenet commented on tests/components/mystrom/test_init.py:
```diff
@@ -0,0 +1,171 @@
+    assert state is None
+
+
+async def test_init_cannot_connect(hass: HomeAssistant) -> None:
```
I think this test should be split in two.
One for failed connect, and one for failed get_state.
```suggestion
async def test_init_cannot_connect(hass: HomeAssistant, config_entry: MockConfigEntry) -> None:
```

epenet commented on tests/components/mystrom/test_init.py:
```diff
@@ -0,0 +1,171 @@
+    ), patch("pymystrom.switch.MyStromSwitch.get_state", return_value={}), patch(
+        "pymystrom.bulb.MyStromBulb.get_state", return_value={}
+    ):
+        config_entry = MockConfigEntry(
+            domain=DOMAIN,
+            entry_id=ENTRY_ID,
+            unique_id="uuid",
+            data={
+                CONF_HOST: "1.1.1.1",
+                CONF_NAME: "test",
+            },
+            title="myStrom",
+        )
+        config_entry.add_to_hass(hass)
+
```
```suggestion
```

epenet commented on tests/components/mystrom/test_init.py:
```diff
@@ -0,0 +1,171 @@
+    ), patch(
+        "pymystrom.bulb.MyStromBulb.get_state", side_effect=MyStromConnectionError()
+    ):
+        config_entry = MockConfigEntry(
+            domain=DOMAIN,
+            entry_id=ENTRY_ID,
+            unique_id="uuid",
+            data={
+                CONF_HOST: "1.1.1.1",
+                CONF_NAME: "test",
+            },
+            title="myStrom",
+        )
+        config_entry.add_to_hass(hass)
+
```
```suggestion
```

pail23 committed:
Update homeassistant/components/mystrom/__init__.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

pail23 committed:
Update homeassistant/components/mystrom/__init__.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

pail23 committed:
Update homeassistant/components/mystrom/config_flow.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

pail23 committed:
Update homeassistant/components/mystrom/config_flow.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

pail23 committed:
Update homeassistant/components/mystrom/config_flow.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

pail23 committed:
Update tests/components/mystrom/test_init.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

pail23 committed:
Update tests/components/mystrom/test_init.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

pail23 committed:
Update tests/components/mystrom/test_init.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

pail23 committed:
Update tests/components/mystrom/test_init.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

pail23 committed:
Update homeassistant/components/mystrom/light.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

pail23 committed:
Update homeassistant/components/mystrom/switch.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

pail23 committed:
fix review comments

epenet commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,134 @@
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"] == {}
```
Add an insert for the step here also please.

epenet commented on tests/components/mystrom/test_init.py:
```diff
@@ -0,0 +1,139 @@
+) -> None:
+    """Test the initialization of a myStrom switch."""
+    await init_integration(hass, config_entry, 101)
+    state = hass.states.get("switch." + DEVICE_NAME)
```
I suggest that you hard-code it here.
```suggestion
    state = hass.states.get(entity_id:="switch.xxx")
```

pail23 replied:
Changed, but I had to change the unload part of the test. Therefor the entity_id is not required

epenet commented on tests/components/mystrom/test_init.py:
```diff
@@ -0,0 +1,139 @@
+async def test_init_bulb(hass: HomeAssistant, config_entry: MockConfigEntry) -> None:
+    """Test the initialization of a myStrom bulb."""
+    await init_integration(hass, config_entry, 102)
+    state = hass.states.get("light." + DEVICE_NAME)
```
I suggest that you hard-code it here.
```suggestion
    state = hass.states.get("light.xxx")
```

pail23 replied:
done

epenet commented on tests/components/mystrom/test_init.py:
```diff
@@ -0,0 +1,139 @@
+    assert state is not None
+    await hass.config_entries.async_unload(config_entry.entry_id)
+    await hass.async_block_till_done()
+    state = hass.states.get("switch.test")
```
```suggestion
    state = hass.states.get(entity_id)
```

pail23 replied:
changed to a differenttest criteria

epenet commented on tests/components/mystrom/test_init.py:
```diff
@@ -0,0 +1,139 @@
+    assert config_entry.state == ConfigEntryState.SETUP_ERROR
+
+
+async def test_unload(hass: HomeAssistant, config_entry: MockConfigEntry) -> None:
+    """Test the unloading of a myStrom witch."""
+    await init_integration(hass, config_entry, 101)
+    await hass.config_entries.async_unload(config_entry.entry_id)
+    await hass.async_block_till_done()
+    state = hass.states.get("switch.test")
+    assert state is None
```
```suggestion
```

pail23 replied:
removed

pail23 committed:
fix review comments

pail23 commented on tests/components/mystrom/conftest.py:
```diff
@@ -0,0 +1,38 @@
+from tests.common import MockConfigEntry
+
+ENTRY_ID = "uuid"
+DEVICE_NAME = "testmyStromdevice"
```
done

pail23 commented on tests/components/mystrom/test_init.py:
```diff
@@ -0,0 +1,139 @@
+from homeassistant.config_entries import ConfigEntryState
+from homeassistant.core import HomeAssistant
+
+from .conftest import DEVICE_NAME
```
done

pail23 commented on tests/components/mystrom/test_init.py:
```diff
@@ -0,0 +1,139 @@
+        await hass.config_entries.async_setup(config_entry.entry_id)
+        await hass.async_block_till_done()
+
+        assert config_entry.state == ConfigEntryState.SETUP_ERROR
```
done

pail23 commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,134 @@
+    entry = MockConfigEntry(
+        domain=DOMAIN, data={CONF_HOST: "1.1.1.1"}, unique_id=DEVICE_MAC
+    )
+    entry.add_to_hass(hass)
```
done

pail23 committed:
Merge branch 'dev' into mystrom_config

pail23 commented on tests/components/mystrom/conftest.py:
```diff
@@ -0,0 +1,38 @@
+    config_entry = MockConfigEntry(
+        domain=DOMAIN,
+        entry_id=ENTRY_ID,
+        unique_id="uuid",
```
done

epenet commented on homeassistant/components/mystrom/strings.json:
```diff
@@ -0,0 +1,23 @@
+    "step": {
+      "user": {
+        "data": {
+          "host": "[%key:common::config_flow::data::host%]"
```
This is now missing a name entry.

pail23 replied:
Added the name entry

epenet commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,131 @@
+
+from tests.common import MockConfigEntry
+
+DEVICE_MAC = "6001940376EB"
+DEVICE_NAME = "myStrom Device"
+
```
You can import `DEVICE_MAC` from `conftest` module.
You shouldn't need to declare `DEVICE_NAME`

epenet commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,131 @@
+                "host": "1.1.1.1",
+            },
+        )
+        assert result2["title"] == DEVICE_NAME
```
```suggestion
        assert result2["title"] == "myStrom Device"
```

epenet replied:
```suggestion
        assert result2["title"] == "myStrom Device"
```

epenet commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,131 @@
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import FlowResultType
+
+from tests.common import MockConfigEntry
+
+DEVICE_MAC = "6001940376EB"
+DEVICE_NAME = "myStrom Device"
```
```suggestion
from .conftest import DEVICE_MAC

from tests.common import MockConfigEntry
```

epenet reviewed (changes requested):
A few remaining comments

epenet reviewed (changes requested):
Last couple of comments have not been adressed.

pail23 committed:
Update tests/components/mystrom/test_config_flow.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

pail23 committed:
Update tests/components/mystrom/test_config_flow.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

pail23 committed:
fix review comment

pail23 committed:
Merge branch 'dev' into mystrom_config

pail23 commented:
@epenet I hope now I catched all your comments

epenet commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,130 @@
+        assert result2["type"] == FlowResultType.ABORT
+        assert result2["reason"] == "already_configured"
+
+    assert len(mock_session.mock_calls) == 1
```
```suggestion
    mock_session.assert_called_once()
```

epenet commented on tests/components/mystrom/test_config_flow.py:
```diff
@@ -0,0 +1,130 @@
+                "host": "1.1.1.1",
+            },
+        )
+        assert result2["title"] == "myStrom Device"
+        assert result2["data"] == {"host": "1.1.1.1"}
+        assert result2["type"] == FlowResultType.CREATE_ENTRY
```
Missing `block_till_done`
```suggestion
        await hass.async_block_till_done()
    assert result2["type"] == FlowResultType.CREATE_ENTRY
    assert result2["title"] == "myStrom Device"
    assert result2["data"] == {"host": "1.1.1.1"}
```

epenet reviewed (changes requested):
Last set of changes, then good to go.

pail23 committed:
Update tests/components/mystrom/test_config_flow.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

pail23 committed:
Update tests/components/mystrom/test_config_flow.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

pail23 committed:
Update tests/components/mystrom/test_config_flow.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

pail23 committed:
Update tests/components/mystrom/test_config_flow.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

pail23 committed:
Update tests/components/mystrom/test_config_flow.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

pail23 committed:
Update tests/components/mystrom/test_config_flow.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

pail23 commented:
I have merged your proposals. Many thanks for your help!

epenet reviewed (approved):
LGTM

epenet commented:
Aaarrgh ! just realised this is missing a documentation PR.
@pail23 can you please open the corresponding doc PR?

pail23 commented:
@epenet I have created home-assistant/home-assistant.io#27667
